### ISSUE TITLE
Node6 incompatibilities with `querystrings`

### ISSUE THREAD

**Issue by cmidkiff87** [js-data/js-data#571]
Node 6 changed the [`querystring.parse`](https://nodejs.org/docs/latest-v6.x/api/querystring.html#querystring_querystring_parse_str_sep_eq_options) to return an object whose `__proto__` property [no longer points back to `Object`](https://github.com/nodejs/node/pull/6055).  This was done to allow for more esoteric query string names, such as `https://example.com/route?__proto__=some-value`, which would conflict with the native js object properties.

However, this causes js-data to error when calling [`hasOwnProperty()`](https://github.com/js-data/js-data/blob/v3.0.8/src/utils.js#L373) if you pass it the `querystring` object.  

This is especially problematic when using [js-data-http](https://www.js-data.io/docs/js-data-http) as it automatically sends the requests with the url's query strings in the format js-data expects.  

It can be fixed by using 
```js
Object.prototype.hasOwnProperty(query, propKey);
```
in place of
```js
query.hasOwnProperty(propKey);
```

---

The libraries involved are 
* api:
  * node 5.11.1 // yikes
  * node 6.17.1
  * js-data 3.0.8
  * koa 1.7.0
* webapp
  * js-data 3.0.0-rc.4
  * js-data-http 3.0.0-rc.2

---

It can also be fixed externally by shimming the `hasOwnProperty()` function onto the querystring
```js
exports.jsData = function() {
  return function*(next) {
    this.jsData = {};

    // `this.query` is the result of the `querystring.parse()` function
    this.jsData.query = this.query; 
    
    function hasOwnPropertyShim(key) {
      // `this` === `this.jsData.query`
      return Object.hasOwnProperty.call(this, key);
    }

    // make the prop unenumerable, or it will get included in the sql query
    Object.defineProperty(this.jsData.query, 'hasOwnProperty', {
      value: hasOwnPropertyShim,
      enumerable: false,
    });

    // ...
  }
}
```
However, that solution isn't super intuitive without some digging

**Comment by zuzusik** [comment]
Node 6 has been out of support for quite a time now. Is it also reproducible in newer versions of Node?

**Comment by cmidkiff87** [comment]
Yes, I can't remember exactly which versions I've tried it with, but the [latest (14.x) docs](https://nodejs.org/docs/latest-v14.x/api/querystring.html#querystring_querystring_parse_str_sep_eq_options) state that 
> The object returned by the `querystring.parse()` method does not prototypically inherit from the JavaScript `Object`. This means that typical `Object` methods such as `obj.toString()`, `obj.hasOwnProperty()`, and others are not defined and will not work.

Of course, this is only an issue when passing the parsed `querystring` object directly to js-data (as we have), and is "fixable" by shimming the object with the required functions.

**Comment by crobinson42** [comment]
I do this on server side middleware for requests:

// Patch request query to be an object for operating on it as an Object
// this is because node.js strips down the prototype for performance/speed
(ctx, next) => {
  if (ctx.request.query && typeof ctx.request.query === 'object') {
    // node.js removed the Object.prototype on querystring https://github.com/nodejs/node/pull/6055
    // coerce query to inherit from Object.prototype
    const def = {
      value: { ...ctx.request.query },
      writable: false,
    }
    Object.defineProperty(ctx.request, 'query', def)
    Object.defineProperty(ctx, 'query', def)
  }

  return next()
},

**Comment by cmidkiff87** [comment]
Then perhaps there should be a note about this somewhere in the documentation, rather than updating js-data?

(since this is probably more of an issue for upgrading legacy projects than it is for new projects)

**Comment by crobinson42** [comment]
I’m not sure that I see this is anything specific to js-data itself, it’s a Node.js quark.

**Comment by zuzusik** [comment]
I think it might be handy to add this as some utility to js-data, which can
be optionally called if required and print some warnings in console when it
comes to the described situation with the advice to run this utility

пн, 13 лип. 2020 о 12:55 cmidkiff87 <notifications@github.com> пише:
